File: articles/personality-wellbeing-and-mental-health/c/wellbeing_summary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 4096

int main(void) {
    FILE *fp = fopen("../data/synthetic_personality_wellbeing_mental_health.csv", "r");
    if (!fp) {
        perror("Could not open data file");
        return 1;
    }

    char line[MAX_LINE];
    if (!fgets(line, sizeof(line), fp)) {
        fprintf(stderr, "Missing header\n");
        fclose(fp);
        return 1;
    }

    int n = 0;
    double wellbeing_sum = 0.0;

    while (fgets(line, sizeof(line), fp)) {
        char *token;
        int col = 0;
        token = strtok(line, ",");
        while (token != NULL) {
            if (col == 15) {
                wellbeing_sum += atof(token);
            }
            token = strtok(NULL, ",");
            col++;
        }
        n++;
    }

    fclose(fp);

    if (n == 0) {
        fprintf(stderr, "No rows\n");
        return 1;
    }

    FILE *out = fopen("../outputs/c_wellbeing_summary.txt", "w");
    if (!out) {
        perror("Could not open output file");
        return 1;
    }

    fprintf(out, "n=%d\nwellbeing_score_mean=%.4f\n", n, wellbeing_sum / n);
    fclose(out);

    printf("Wrote C output: ../outputs/c_wellbeing_summary.txt\n");
    return 0;
}

File: articles/personality-wellbeing-and-mental-health/fortran/wellbeing_summary.f90
program wellbeing_summary
  implicit none

  character(len=4096) :: line
  character(len=256) :: field
  integer :: unit, ios, n, pos, next_pos, col
  real :: wellbeing, wellbeing_sum

  wellbeing_sum = 0.0
  n = 0

  open(newunit=unit, file="../data/synthetic_personality_wellbeing_mental_health.csv", status="old", action="read", iostat=ios)
  if (ios /= 0) then
    print *, "Could not open data file"
    stop 1
  end if

  read(unit, '(A)', iostat=ios) line

  do
    read(unit, '(A)', iostat=ios) line
    if (ios /= 0) exit

    pos = 1
    col = 1
    wellbeing = 0.0

    do
      next_pos = index(line(pos:), ",")
      if (next_pos == 0) then
        field = adjustl(line(pos:))
      else
        field = adjustl(line(pos:pos+next_pos-2))
      end if

      if (col == 16) read(field, *) wellbeing

      if (next_pos == 0) exit
      pos = pos + next_pos
      col = col + 1
    end do

    wellbeing_sum = wellbeing_sum + wellbeing
    n = n + 1
  end do

  close(unit)

  open(newunit=unit, file="../outputs/fortran_wellbeing_summary.txt", status="replace", action="write")
  write(unit, '(A,I0)') "n=", n
  write(unit, '(A,F8.4)') "wellbeing_score_mean=", wellbeing_sum / real(n)
  close(unit)

  print *, "Wrote Fortran output: ../outputs/fortran_wellbeing_summary.txt"
end program wellbeing_summary
